File: Experiment-1/spectral_norm/spectral_norm.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <cstdlib>
#include <iomanip>

using namespace std;

double eval_A(int i, int j, int N) {
    return 1.0 / ((i + j) * (i + j + 1) / 2 + i + 1);
}

void eval_A_times_u(int N, const vector<double>& u, vector<double>& Au) {
    for (int i = 0; i < N; i++) {
        Au[i] = 0;
        for (int j = 0; j < N; j++) Au[i] += eval_A(i, j, N) * u[j];
    }
}

void eval_At_times_u(int N, const vector<double>& u, vector<double>& Au) {
    for (int i = 0; i < N; i++) {
        Au[i] = 0;
        for (int j = 0; j < N; j++) Au[i] += eval_A(j, i, N) * u[j];
    }
}

void eval_AtA_times_u(int N, const vector<double>& u, vector<double>& AtAu) {
    vector<double> v(N);
    eval_A_times_u(N, u, v);
    eval_At_times_u(N, v, AtAu);
}

int main(int argc, char *argv[]) {
    int N = (argc > 1) ? atoi(argv[1]) : 260; // Default size of the problem
    int LOOPS = (argc > 2) ? atoi(argv[2]) : 100; // Default number of iterations
    vector<double> u(N, 1.0), v(N);
    double total = 0.0;

    for (int loop = 0; loop < LOOPS; loop++) {
        fill(u.begin(), u.end(), 1);

        for (int i = 0; i < 10; i++) {
            eval_AtA_times_u(N, u, v);
            eval_AtA_times_u(N, v, u);
        }

        double vBv = 0, vv = 0;
        for (int i = 0; i < N; i++) {
            vBv += u[i] * v[i];
            vv += v[i] * v[i];
        }
        total += vBv + vv;
    }

    cout << fixed << setprecision(3) << total << endl;
    return 0;
}

File: Experiment-1/spectral_norm/spectral_norm.py
# """
# MathWorld: "Hundred-Dollar, Hundred-Digit Challenge Problems", Challenge #3.
# http://mathworld.wolfram.com/Hundred-DollarHundred-DigitChallengeProblems.html

# The Computer Language Benchmarks Game
# http://benchmarksgame.alioth.debian.org/u64q/spectralnorm-description.html#spectralnorm

# Contributed by Sebastien Loisel
# Fixed by Isaac Gouy
# Sped up by Josh Goldfoot
# Dirtily sped up by Simon Descarpentries
# Concurrency by Jason Stitt
# Adapted for Codon by @arshajii
# """

# from time import time

# DEFAULT_N = 260

# def eval_A(i, j):
#     return 1.0 / ((i + j) * (i + j + 1) // 2 + i + 1)

# def eval_times_u(func, u):
#     return [func((i, u)) for i in range(len(list(u)))]

# def part_At_times_u(i_u):
#     i, u = i_u
#     partial_sum = 0.
#     for j, u_j in enumerate(u):
#         partial_sum += eval_A(j, i) * u_j
#     return partial_sum

# def part_A_times_u(i_u):
#     i, u = i_u
#     partial_sum = 0.
#     for j, u_j in enumerate(u):
#         partial_sum += eval_A(i, j) * u_j
#     return partial_sum

# def eval_AtA_times_u(u):
#     return eval_times_u(part_At_times_u, eval_times_u(part_A_times_u, u))

# def bench_spectral_norm(loops):
#     range_it = range(loops)
#     total = 0.
#     for _ in range_it:
#         u = [1.] * DEFAULT_N
#         v = None
#         for dummy in range(10):
#             v = eval_AtA_times_u(u)
#             u = eval_AtA_times_u(v)
#         vBv = vv = 0.
#         for ue, ve in zip(u, v):
#             vBv += ue * ve
#             vv += ve * ve
#         total += vBv + vv
#     return total

# # t0 = time()
# print(bench_spectral_norm(100))
# # t1 = time()
# # print(t1 - t0)
import sys
from time import time

def eval_A(i, j):
    return 1.0 / ((i + j) * (i + j + 1) // 2 + i + 1)

def eval_times_u(func, u, N):
    return [func((i, u, N)) for i in range(N)]

def part_At_times_u(i_u_N):
    i, u, N = i_u_N
    partial_sum = 0.
    for j, u_j in enumerate(u):
        partial_sum += eval_A(j, i) * u_j
    return partial_sum

def part_A_times_u(i_u_N):
    i, u, N = i_u_N
    partial_sum = 0.
    for j, u_j in enumerate(u):
        partial_sum += eval_A(i, j) * u_j
    return partial_sum

def eval_AtA_times_u(u, N):
    return eval_times_u(part_At_times_u, eval_times_u(part_A_times_u, u, N), N)

def bench_spectral_norm(N, loops):
    total = 0.
    for _ in range(loops):
        u = [1.] * N
        for dummy in range(10):
            v = eval_AtA_times_u(u, N)
            u = eval_AtA_times_u(v, N)
        vBv = vv = 0.
        for ue, ve in zip(u, v):
            vBv += ue * ve
            vv += ve * ve
        total += vBv + vv
    return total

if __name__ == "__main__":
    N = int(sys.argv[1]) if len(sys.argv) > 1 else 260
    LOOPS = int(sys.argv[2]) if len(sys.argv) > 2 else 100
    print(f"{bench_spectral_norm(N, LOOPS):.3f}")

